### PR TITLE
ESP32-C3 integration

### PR DESCRIPTION
**Pull request by unlogisch04**
Draft Pull request for discussions and feedback's.

Todo:
- I2CScan need to be changed to work with ESP32C3. Missing here a Compiler flag to be sure that we have a ESP32C3.
- Board the firebeetle32c3 https://github.com/platformio/platform-espressif32/pull/829
- Testing on other boards.

Testing:
- [x] Testing ICM20948
- [x] Testing MPU6050
- Testing MPU9250 (i can not test that one)
- Testing BNO085 (i can not test that one)
- [x] Testing BMI160 (BMI160 seems to crash after some console "REBOOT" or after a short time the data seems only from magnetometer, not sure if it is only my BMI160 bad or a problem with the lib not correctly init.)
- [ ] Testing Server WiFi Window (does not work filter need to be changed to allow "Serielles USB-Gerät" )
-> does not work
- [x] Board dfrobot_beetle_esp32c3 (need atm custom board)
- [x] Board esp32-c3-devkitm-1
- [x] Board lolin_c3_mini
- [x] Board ESP8266
- [x] Board ESP32

### CODE REVIEW

**Comment by unlogisch04** [comment]
With the firebeetle32c3 i can not connect with the WiFi Window to the server.
It reports as: "Serielles USB-Gerät (COM10)" on Windows. This will not trigger the filter of the Server.

**Review by TheButlah — COMMENTED** [review]
Tested on my `ferrous_slime` esp32c3 board, and a d1 mini. Both with MPU6050. It worked, with the following minor tweaks:

- On the esp32c3, I lost most of the initial serial printouts done in the setup until I added a `delay(2000)` right after the serial was initialized. I think we should add this delay in on the esp32c3 build behind an `#ifdef`
- The firebeetle32c3 board doesn't exist, I assume @unlogisch04 must have added his own custom board description json. To get this mergeable, we should just have two options probably - the dev kit and the lolin_c3_mini. This whole esp32c3 section will then be commented out and users can pick which one.
- We should add a comment in the esp32c3 section of the `platformio.ini` that explains which board to pick - if your board uses USB <-> UART chip, pick the `esp32-c3-devkitm-1` board. If it uses the onboard USB serial directly (D+ and D- are connected to GPIO19 and GPIO18 directly), then use the `lolin_c3_mini`. If using the latter, you will need to push the Flash/boot button, or otherwise set the appropriate strapping pins. Link them to this: https://docs.espressif.com/projects/esptool/en/latest/esp32c3/advanced-topics/boot-mode-selection.html

**Comment by unlogisch04** [comment]
I request to review and test. I did not squash till now, so changes can be tracked better.
@TheDevMinerTV @Eirenliel 
Is it ok to add 2 boards to the CI?

**Comment by TheDevMinerTV** [comment]
we're not paying for CI minutes, so yes.
note to self: I have to parallelize the building at some point though.

**Comment by unlogisch04** [comment]
It should be ready for review.
Only open point is, that the the Server: https://github.com/SlimeVR/SlimeVR-Server/issues/108#issuecomment-1208546510

**Comment by TheButlah** [comment]
Will review maybe in a week when I get back home

File: README.md [hunk 1]
```diff
@@ -48,6 +48,13 @@ Firmware can work with both ESP8266 and ESP32. Please edit `defines.h` and set y
   * LED will turn off when calibration is complete
   * You don't have to calibrate next time you power it off, calibration values will be saved for the next use
 
+## Infos about ESP32-C3 with direct connection to USB
+The ESP32-C3 has 2 ways to connect the serial port. One is directly via the onboard USB CDC or via the onboard UART.
+When the Chip is connected to the USB-CDC, the serial port shows is displayed as "USB Serial Port". The SlimeVR server is not currently connecting to this port.
+If you want to set your WiFi credentials, you can use the PlatformIO serial console.
+There you have to enter the following:
+`SET WIFI "SSID" "PASSWORD"`
```
Comment (TheDevMinerTV) [inline README.md:56]:
```suggestion
The ESP32-C3 has two ways to connect the serial port. One is directly via the onboard USB CDC or via the onboard UART.
When the chip is connected to the USB CDC, the serial port shows as `USB Serial Port` in Device Manager. The SlimeVR server will currently not connect to this port.
If you want to set your WiFi credentials, you can use the PlatformIO serial console.
There you have to enter the following: `SET WIFI "SSID" "PASSWORD"`
```

File: lib/i2cscan/i2cscan.cpp [hunk 2]
```diff
@@ -1,11 +1,18 @@
 #include "i2cscan.h"
+#include "../../src/globals.h"
 
 #ifdef ESP8266
 uint8_t portArray[] = {16, 5, 4, 2, 14, 12, 13};
+uint8_t portExclude[] = {LED_PIN};
 String portMap[] = {"D0", "D1", "D2", "D4", "D5", "D6", "D7"};
+#elif defined(ESP32C3)  //ESP32C3 has not that much ports as a ESP32
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:8]:
```suggestion
// ESP32C3 has not as many ports as the ESP32
#elif defined(ESP32C3)
```

File: lib/i2cscan/i2cscan.cpp [hunk 3]
```diff
@@ -1,11 +1,18 @@
 #include "i2cscan.h"
+#include "../../src/globals.h"
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:2]: Do you really need this include?

File: lib/i2cscan/i2cscan.cpp [hunk 4]
```diff
@@ -43,12 +50,45 @@ namespace I2CSCAN
...
+        }
+        //Serial.printf("InArray: %d, %d, %d, result: %x\n", search, search2, size, ret);
+        return ret;
+    }
+    
     bool checkI2C(uint8_t i, uint8_t j)
     {
         bool found = false;
-        Wire.begin(portArray[i], portArray[j]);
+#if ESP32
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:88]: Shouldn't this be for the ESP32C3?

File: lib/i2cscan/i2cscan.cpp [hunk 5]
```diff
@@ -43,12 +50,45 @@ namespace I2CSCAN
         if(!found) {
             Serial.println("[ERR] I2C: No I2C devices found");
         }
+    // begin Reset the Wire interface back to the original Values
+    #if ESP32
+        Wire.end();
+    #endif
+        Wire.begin(static_cast<int>(PIN_IMU_SDA), static_cast<int>(PIN_IMU_SCL));
+    // end Reset the Wire interface back to the original Values
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:58]:
```suggestion
#if ESP32
        Wire.end();
#endif

        // Reset the I2C interface back to it's original values
        Wire.begin(static_cast<int>(PIN_IMU_SDA), static_cast<int>(PIN_IMU_SCL));
```

File: lib/i2cscan/i2cscan.cpp [hunk 6]
```diff
@@ -43,12 +50,45 @@ namespace I2CSCAN
...
+                if (search2 == arraysearch[k]) 
+                {
+                    ret = true;
+                    k = size;
+                }
+            }
+        }
+        //Serial.printf("InArray: %d, %d, %d, result: %x\n", search, search2, size, ret);
+        return ret;
+    }
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:83]:
```suggestion
    boolean inArray(uint8_t value, uint8_t* array, size_t arraySize)
    {
        for (size_t i = 0; i < arraySize; i++)
        {
            if (value == array[i]) 
            {
                return true;
            }
        }

        return false;
    }
```

File: lib/i2cscan/i2cscan.cpp [hunk 7]
```diff
@@ -33,7 +40,7 @@ namespace I2CSCAN
         {
             for (uint8_t j = 0; j < sizeof(portArray); j++)
             {
-                if (i != j)
+                if ((i != j) && !inArray(portArray[i], portArray[j], portExclude, sizeof(portExclude)))
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:43]:
```suggestion
                if ((i != j) && !inArray(portArray[i], portExclude, sizeof(portExclude)) && !inArray(portArray[j], portExclude, sizeof(portExclude)))
```

File: lib/i2cscan/i2cscan.h [hunk 8]
```diff
@@ -10,6 +10,7 @@ namespace I2CSCAN {
     bool isI2CExist(uint8_t addr);
     uint8_t pickDevice(uint8_t addr1, uint8_t addr2, bool scanIfNotFound);
     int clearBus(uint8_t SDA, uint8_t SCL);
+    boolean inArray(uint8_t search, uint8_t search2, uint8_t arraysearch[], uint8_t size);
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.h:13]:
```suggestion
    boolean inArray(uint8_t value, uint8_t* arr, size_t arrSize);
```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

File: ci/devices.json [hunk 9]
```diff
@@ -5,13 +5,18 @@
     "board": "SLIMEVR"
   },
   {
-    "platform": "espressif32 @ 4.4.0",
+    "platform": "espressif32 @ 5.1.1",
```
[the remark this answers is not in the document]
Comment (unlogisch04) [reply ci/devices.json:8]: Forgot thx.

File: lib/i2cscan/i2cscan.cpp (hunk 3 above)
Comment (unlogisch04) [reply lib/i2cscan/i2cscan.cpp:2]: yes, it is needed for `uint8_t portExclude[] = {LED_PIN};`

File: lib/i2cscan/i2cscan.cpp (hunk 4 above)
Comment (unlogisch04) [reply lib/i2cscan/i2cscan.cpp:89]: No, it is for ESP32 (espressif32). They did add a check on the new versions (i think from 5.1 on)

File: lib/i2cscan/i2cscan.cpp (hunk 6 above)
Comment (unlogisch04) [reply lib/i2cscan/i2cscan.cpp:84]: Now we have to loop 2 times trough the same array. But agree it looks better.

File: ci/devices.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (unlogisch04) [reply ci/devices.json:8]: fixed with commit [ff879a8](https://github.com/SlimeVR/SlimeVR-Tracker-ESP/pull/178/commits/ff879a8dd88a51f2addd4c0ae69620154ed2fdb8)
also corrected not needed compile flags (new in board definition)

**Review by Eirenliel — APPROVED** [review]

File: lib/i2cscan/i2cscan.cpp [hunk 10]
```diff
@@ -43,12 +51,34 @@ namespace I2CSCAN
...
         }
+#if ESP32
+        Wire.end();
+#endif
+
+        // Reset the I2C interface back to it's original values
+        Wire.begin(static_cast<int>(PIN_IMU_SDA), static_cast<int>(PIN_IMU_SCL));
     }
 
+    boolean inArray(uint8_t value, uint8_t* array, size_t arraySize)
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:62]:
```suggestion
    bool inArray(uint8_t value, uint8_t* array, size_t arraySize)
```

File: lib/i2cscan/i2cscan.cpp [hunk 11]
```diff
@@ -43,12 +51,34 @@ namespace I2CSCAN
...
+    }
+    
     bool checkI2C(uint8_t i, uint8_t j)
     {
         bool found = false;
-        Wire.begin(portArray[i], portArray[j]);
+#if ESP32
+        Wire.end();
+#endif
+        Wire.begin((int)portArray[i], (int)portArray[j]);
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:81]:
```suggestion

#if ESP32
        Wire.end();
#endif

        Wire.begin((int)portArray[i], (int)portArray[j]);

```

File: lib/i2cscan/i2cscan.cpp [hunk 12]
```diff
@@ -43,12 +51,34 @@ namespace I2CSCAN
         if(!found) {
             Serial.println("[ERR] I2C: No I2C devices found");
         }
+#if ESP32
+        Wire.end();
+#endif
```
Comment (TheDevMinerTV) [inline lib/i2cscan/i2cscan.cpp:56]:
```suggestion

#if ESP32
        Wire.end();
#endif
```

File: ci/build.py [hunk 13]
```diff
@@ -25,10 +27,19 @@ def __init__(self,  platform: str, board: Board, platformio_board: str) -> None:
...
+
+        if self.board == Board.LOLIN_C3_MINI:
+            section += "build_flags = \n"
+            section += " ${env.build_flags}\n"
+            section += " -DESP32C3\n"
+
+        if self.board == Board.ES32C3DEVKITM1:
+            section += "build_flags = \n"
+            section += " ${env.build_flags}\n"
+            section += " -DESP32C3\n"
```
Comment (TheDevMinerTV) [inline ci/build.py:42]:
```suggestion
            section += " -DESP32C3\n"

```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

**Review by TheDevMinerTV — APPROVED** [review]

**Pull request merged**